haskell/test/SMT.hs

```diff
@@ -0,0 +1,134 @@
+
+module SMT
+  ( hprop_SMT
+  ) where
+
+import Universum hiding (drop, swap)
+
+import Data.Sequence (Seq(..))
+import Hedgehog
+
+import Lorentz hiding (now, (>>))
+import Michelson.Typed.Convert (convertContract, untypeValue)
+import Morley.Nettest
+import qualified Morley.Micheline as MM
+import Morley.Nettest.Pure
+
+import Ligo.BaseDAO.Contract
+import Ligo.BaseDAO.Types
+import Test.Ligo.BaseDAO.Common.StorageHelper as StorageHelper
+import Test.Ligo.BaseDAO.Common
+import SMT.Model
+import SMT.Model.Types
+import SMT.Gen
+import SMT.Types
+
+hprop_SMT :: Property
+hprop_SMT =
+  withTests 20 $ property $ do
+    PropertyTestInput (contractCalls, ms) <- forAll genPropertyTestInput
+    nettestTestProp $
+      (uncapsNettestEmulated $ do
+          -- Set initial level
+          advanceLevel (ms & msLevel)
+
+          guardianContract <- originateSimple "guardian" () dummyGuardianContract
+          tokenContract <- originateSimple "TokenContract" [] dummyFA2Contract
+
+          let newMs = addAddresses ms (unTAddress guardianContract) (unTAddress tokenContract)
+
+          -- Originate Dao for nettest
+          let
+            originateData = UntypedOriginateData
+              { uodName = "BaseDAO"
+              , uodBalance = toMutez 0
+              , uodStorage = untypeValue $ toVal $ (newMs & msFullStorage)
+              , uodContract = convertContract baseDAOContractLigo
+              }
+
+          daoUntyped <- originateUntyped originateData
+          let dao = TAddress @Parameter daoUntyped
+
+          callEntrypointAndCompare dao contractCalls newMs
+
+      )
+      emulatedImplIntegrational
+
+  where
+    -- Reason: guardian address and governance token address are not known until we originate the contract in nettest
+    addAddresses ms guardAddr govAddr = ms
+        { msFullStorage = (ms & msFullStorage)
+            { fsStorage = (ms & msFullStorage & fsStorage)
+                { sGuardian = guardAddr
+                , sGovernanceToken = (ms & msFullStorage & fsStorage & sGovernanceToken)
+                    { gtAddress = govAddr
+                    }
+                }
+            }
+        }
+
+callEntrypointAndCompare :: TAddress Parameter -> [BaseDaoContractCall] -> ModelState -> EmulatedT PureM ()
+callEntrypointAndCompare _ [] _ = pure ()
+callEntrypointAndCompare dao (cc:ccs) ms = do
+
+  let (haskellResult, updatedMs) = runBaseDaoHaskell cc ms
+  (ligoResult, ligoFs) <- runBaseDaoLigo dao cc
+
+  -- TODO: removed
+  -- let !_ = trace (pretty (cc)) ()
+  -- let !_ = trace (pretty (updatedMs & msFullStorage & fsStorage & sQuorumThresholdAtCycle)) ()
+  -- let !_ = trace (pretty (updatedMs & msFullStorage & fsStorage & sFreezeHistory)) ()
+  -- let !_ = trace (pretty (ligoFs & fsStorage & sQuorumThresholdAtCycle)) ()
+  -- let !_ = trace (pretty (ligoFs & fsStorage & sFreezeHistory)) ()
+
+  -- Note: Comparing `cc` is useful since it appears in the log, when something went wrong.
```

Comment:
Can't we compare the stuff manually and throw a custom test failure with a message with what ever details we find interesting?

Reply:
I've added `printResult`, which should throw better errors.